packages/seal/src/api.ts

```diff
@@ -0,0 +1,262 @@
+export const MAX_U8 = 255;
+
+const IBEEncryptions = bcs.enum('IBEEncryptions', {
+	BonehFranklinBLS12381: bcs.struct('BonehFranklinBLS12381', {
+		encapsulation: bcs.bytes(96),
+		shares: bcs.vector(bcs.bytes(32)),
+	}),
+});
+export type IBEEncryptionsType = typeof IBEEncryptions.$inferType;
+
+/**
+ * The encrypted object format. Should be aligned with the Rust implementation.
+ */
+export const EncryptedObject = bcs.struct('EncryptedObject', {
+	version: bcs.U8,
+	package_id: bcs.bytes(32),
+	inner_id: bcs.vector(bcs.U8),
+	services: bcs.vector(bcs.tuple([bcs.bytes(32), bcs.U8])),
+	threshold: bcs.U8,
+	encrypted_shares: IBEEncryptions,
+	ciphertext: Ciphertext,
+});
+
+export type VerifiedKeyServerInfo = {
+	id: string;
+	pk: Uint8Array;
+};
+
+/**
+ * Query all register service object by its type. Visits the registered URL's /service endpoint and verified
+ * the returned pop and public key, then return the object id and public key.
+ *
+ * @param keyServerObjectId - The key server object ID.
+ * @returns - An array of [service_object_id, pk_bytes].
+ */
+export async function fetchVerifiedKeyServerInfo({
+	keyServerObjectId,
+	network = 'mainnet',
+	client = new SuiClient({ url: getFullnodeUrl(network) }),
+}: {
+	keyServerObjectId: string;
+	network?: 'testnet' | 'mainnet';
+	client?: SuiClient;
+}): Promise<VerifiedKeyServerInfo> {
+	return await client
+		.getObject({
+			id: keyServerObjectId,
+			options: {
+				showContent: true,
+			},
+		})
+		.then(async (res) => {
+			if (!res?.data) throw new Error('No data returned');
+			const fields = (res?.data?.content as { fields: { [k: string]: any } })?.fields;
+			const response = await fetch(fields?.url! + '/v1/service', {
+				method: 'GET',
+				headers: {
+					'Content-Type': 'application/json',
+				},
+			});
+			const serviceResponse = await response.json();
+
+			if (serviceResponse.service_id !== keyServerObjectId) {
+				throw new Error('Service id mismatch');
+			}
+			const pk = new Uint8Array(fields?.pk?.fields?.pos0.fields.bytes);
+			// console.log('pk g2', pk.length);
+			// console.log('pop g1', fromB64(serviceResponse.pop).length);
+
+			// let fullMsg = new Uint8Array([...DST_POP, ...pk, ...fromHEX(id)]);
+			// console.log('fullMsg', fullMsg);
+			// let hashedG1 = G1Element.hashToCurve(fullMsg).toBytes();
+
+			// console.log('g1',hashedG1.length);
+			// if (!await bls12_381.verify(fromB64(serviceResponse.pop), hashedG1, pk)) {
+			//   throw new Error("Invalid pop");
+			// }
+			return {
+				id: keyServerObjectId,
+				pk,
+			};
+		});
+}
+
+/** Look up urls of key servers and fetch key from servers with request signature, cert and ephPk, then decrypt the response with ephSk.
+ *
+ * @param serviceIds - The object id's of the key servers.
+ * @param requestSig - The request signature.
+ * @param ptb - The ptb.
+ * @param cert - The certificate.
+ * @param ephSk - The ephemeral public key.
+ * @returns - The secret keys.
+ */
+export async function fetchSks({
```

Comment:
the current code will fail if one use 1 out of 2 and one of the services is down, no